node/src/coordinator.rs

```diff
@@ -317,29 +342,155 @@ impl Coordinator {
         let (network_client, mut channel_receiver, _handle) =
             run_network_client(Arc::new(sender), Box::new(receiver));
 
-        let key = EcdsaSignatureProvider::run_key_generation_client(
-            mpc_config,
-            network_client,
-            &mut channel_receiver,
-        )
-        .await?;
-
-        keyshare_storage
-            .store_key(Keyshare::from_legacy(&LegacyRootKeyshareData {
-                epoch: 0,
-                private_share: key.private_share,
-                public_key: key.public_key,
-            }))
-            .await?;
-        let my_public_key = affine_point_to_public_key(key.public_key)?;
-        chain_txn_sender
-            .send(ChainSendTransactionRequest::VotePk(ChainVotePkArgs {
-                public_key: my_public_key,
-            }))
-            .await?;
-
-        // Just halt and wait for the running state.
-        Ok(MpcJobResult::HaltUntilInterrupted)
+        let is_leader = mpc_config.is_leader_for_keygen();
+        let keyshare_storage = Arc::new(keyshare_storage);
+        if !is_leader {
+            'follower: loop {
+                let channel = channel_receiver.recv().await.unwrap();
+                let task_id = channel.task_id();
+                let MpcTaskId::EcdsaTaskId(EcdsaTaskId::KeyGeneration {
+                    key_event: task_key_event_id,
+                }) = task_id
+                else {
+                    tracing::info!(
+                        "Expected Keygeneration task id, received: {:?}; ignoring.",
+                        task_id,
+                    );
+                    continue 'follower;
+                };
+                tracing::info!(
+                                    "Received Keygeneration task id: {:?}; spawning task, waiting on contract change. Our id; {:?}" ,
+                                    task_id,
+                mpc_config.my_participant_id,
+                                );
+                let max_timeout = 120;
+                let key_event_receiver_cloned = key_event_receiver.clone();
+                let mpc_config_cloned = mpc_config.clone();
+                let chain_txn_sender_cloned = chain_txn_sender.clone();
+                let keyshare_storage_cloned = keyshare_storage.clone();
+                //tokio::spawn(async move {
+                // Wait for the contract to confirm this key event
+                let mut contract_event = key_event_receiver_cloned.borrow().clone();
+                let mut n = 0; // one minute max
+                while contract_event.id != task_key_event_id || !contract_event.started {
```

Comment:
I think this should be:

- While contract event not started and `state.generating_key.next_attempt_id == task_key_event_id.attempt_id`: wait for a change on `key_event_receiver_cloned`.
- Once we exit that loop, if we haven't started or if `contract_event.id != task_key_event_id`, `bail!("Not the active key ID")`

That way, we do not actually need to even have a timeout here. Why? Because the contract will naturally progress towards higher attempt IDs and this will exit whenever the appropriate timeout happens.